### PR TITLE
[WIP] refactor loss layers

### PR DESCRIPTION
**Pull request by piiswrong**
@sbodenstein @pluskid @mli @zackchase @sxjscience 

refactored handling of loss layers. ***Output layers will be deprecated going forward. Will add documentations next.

@sbodenstein I'll make a log_softmax layer that can do reduce on any axis

### CODE REVIEW

**Comment by sbodenstein** [comment]
@piiswrong: we have implemented a numerically stable cross-entropy loss layer in CUDA and C++ that we use internally. It returns a loss per example in a batch, making it suitable for weighting examples as you are doing here. It also has support for both sparse and probability targets. Would this be useful for this?

**Comment by piiswrong** [comment]
@sbodenstein Yes we can use that as the internal op in cross_entropy_loss

**Comment by sbodenstein** [comment]
> Can you specify which axis to sum over?

It does have the option for `multi_output` that `SoftmaxOutput` has. Do you want a general spec?

**Comment by sbodenstein** [comment]
Also @piiswrong: we would like to change CTC loss so that it actually returns the loss per example (currently, it throws away all loss info, which is crazy, as its expensive to compute). Now might be a good time to refactor this as well. Can I make a PR for this? Or are there issues breaking backward compatibility?

**Comment by piiswrong** [comment]
Yes we want a general way to specify axis

about CTC, please make a PR and I'll have a look
